Bluetooth: Simplify Kconfig options and interface between the host and controller

jhedberg opened the pull request:
These patches simplify the Kconfig options (getting rid of unnecessary ones) and also remove the bt_hci_read_static_addr() "hack" that existed between the host and the controller.

jhedberg committed:
Bluetooth: Kconfig: Merge BT_HCI_VS_EXT into BT_HCI_VS

The naming of these two options was problematic, since it's both of them
are about vendor extensions, even though one has _EXT in the name and
the other doesn't. Just merge one option into the other. This has a
slight overhead on the controller side of enabling some more vendor
features if BT_HCI_VS is enabled, but that should hopefully be
acceptable.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg committed:
Bluetooth: Kconfig: Get rid of BT_HCI_VS_EVT

Enabling vendor-specific extensions also implies support for vendor
events, so a separate Kconfig option for that is unnecessary.

One small additional thing this requires is the use of the
__maybe_unused annotation, since there's no-longer a single Kconfig
option that the controller hci.c can use to know that the vendor event
helper symbols are needed.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

jhedberg committed:
Bluetooth: Remove bt_read_static_addr() "hack"

This function was used to shortcut HCI for combined host + controller
builds. It doesn't provide much value and adds complexity to the HCI
driver interface, so just remove it. This means vendor-specific HCI
commands is now the only way for the host to access the same
information.

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

Thalley commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -211,9 +211,9 @@ static uint32_t cis_pending_count;
-#if defined(CONFIG_BT_HCI_VS_EVT)
-static uint64_t vs_events_mask = DEFAULT_VS_EVT_MASK;
-#endif /* CONFIG_BT_HCI_VS_EVT */
+#if defined(CONFIG_BT_HCI_VS)
+__maybe_unused static uint64_t vs_events_mask = DEFAULT_VS_EVT_MASK;
+#endif /* CONFIG_BT_HCI_VS */
```
Is there any chance we can remove the guards altogether? 

IIRC if a `static` variable is only by `static` functions and the calls to the functions are guarded by `IS_ENABLED`, then the linker should be able to remove the `static` variables as well. 

Generally `#if defined` is only really necessary when fields inside structs are used. Is this the case here?

jhedberg replied:
> Is there any chance we can remove the guards altogether?
> 
> IIRC if a `static` variable is only by `static` functions and the calls to the functions are guarded by `IS_ENABLED`, then the linker should be able to remove the `static` variables as well.
> 
> Generally `#if defined` is only really necessary when fields inside structs are used. Is this the case here?

You're mostly right. In this case the necessity comes from direction finding vendor events, whose implementation is guarded by #ifdefs and calls to which come from `case` statements which are also guarded by #ifdefs. So I decided to leave these out of the scope of this PR.

jhedberg replied:
Btw, one more push coming once CI shows green, since I decided to mention these in the migration guide as well (at least some of the changes can be seen as API breakage, or at least something that can cause build breakage in downstream trees).

Thalley reviewed (dismissed):
Thank you! I feel it is getting out of hand with all the Kconfigs and `#if defined` we use in the BT subsystem. Recently started some cleanup in LE audio as well that suffer from much of this same complexity.

jhedberg committed:
doc: migration-guide-3.7: Mention Bluetooth HCI related changes

Mention the changes to Bluetooth HCI Kconfig options and one
semi-private API (bt_read_static_addr).

Signed-off-by: Johan Hedberg <johan.hedberg@gmail.com>

Thalley commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -150,6 +150,13 @@ Analog-to-Digital Converter (ADC)
 Bluetooth HCI
 =============
 
+ * The ``BT_HCI_VS_EXT`` Kconfig option was merged into the :kconfig:option:`BT_HCI_VS` Kconfig
+   option.
+ * The ``BT_HCI_VS_EVT`` Kconfig option was removed, since vendor event support is implicit if
+   the :kconfig:option:`BT_HCI_VS` option is enabled.
+ * The bt_read_static_addr() API was removed. This wasn't really a completely public API, but
+   since it was exposed by the public hci_driver.h header file the removal is mentioned here.
```
It's not completely clear to me when things belong in the migration guide vs. release notes. 

These lines here don't seem like a guide to me though, and as such perhaps belong in the release notes instead?

jhedberg replied:
> It's not completely clear to me when things belong in the migration guide vs. release notes.
> 
> These lines here don't seem like a guide to me though, and as such perhaps belong in the release notes instead?

@Thalley the way I think about it is in terms of "will this break someone's application". If the answer is yes, then I go for the migration guide, otherwise the release notes. In principle these are all changes that *could* break an application, although the `bt_read_static_addr()` API is kind of borderline: even though it's in a public header, the only user of it should really just have been the host stack.

jhedberg replied:
@Thalley ah, I think I understand what you mean by "does it guide you". For `BT_HCI_VS_EXT` the guide is "start using `BT_HCI_VS` instead. For `BT_HCI_VS_EVT` and `bt_read_static_addr()` the advice would (I suppose) be "don't use them". Anyway, I don't mind moving these to the release notes as well, but my understanding was that things that may require changes to applications should go to the migration guide. @carlescufi any thoughts?

cvinayak replied:
> things that may require changes to applications should go to the migration guide

I agree with the above, that as a migration guide, an user needs to take necessary action to resolve build errors and to resolve change in runtime behaviors, if any.

Thalley replied:
If that is the case, shouldn't the migration guide still be a subset of the release notes? 

https://docs.zephyrproject.org/latest/releases/index.html#release-notes states
> Changes that require the user to modify their own application to support the new release may be mentioned in the release notes, but the details regarding what needs to be changed are to be detailed in the release’s migration guide.

So release could be considered _what_ and migration guide could be considered _how_?

carlescufi replied:
The wording is a bit confusing. Anything that requires changing the user's code must be in the migration guide. The release notes is about changes in general, and the criteria for what to include there is a bit more loosely defined. I would place this in the migration guide only in this case.

carlescufi replied:
I will send a PR to clarify this.

cvinayak commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -150,6 +150,13 @@ Analog-to-Digital Converter (ADC)
 Bluetooth HCI
 =============
 
+ * The ``BT_HCI_VS_EXT`` Kconfig option was merged into the :kconfig:option:`BT_HCI_VS` Kconfig
+   option.
```
```suggestion
 * The ``BT_HCI_VS_EXT`` Kconfig option is deleted and the feature is now included in the :kconfig:option:`BT_HCI_VS` Kconfig
   option.
```

cvinayak commented on doc/releases/migration-guide-3.7.rst:
```diff
@@ -150,6 +150,13 @@ Analog-to-Digital Converter (ADC)
+   option.
+ * The ``BT_HCI_VS_EVT`` Kconfig option was removed, since vendor event support is implicit if
+   the :kconfig:option:`BT_HCI_VS` option is enabled.
+ * The bt_read_static_addr() API was removed. This wasn't really a completely public API, but
+   since it was exposed by the public hci_driver.h header file the removal is mentioned here.
```
```suggestion
 * The bt_read_static_addr() API was removed. This wasn't really a completely public API, but
   since it was exposed by the public hci_driver.h header file the removal is mentioned here. Enable the :kconfig:option:`BT_HCI_VS` Kconfig
   option instead, and use vendor specific HCI commands API to get the Controller's Bluetooth static address when available.
```

thoh-ot reviewed (changes requested):
Sorry to be that guy, I just need a day to verify that this doesn't give issues with our controller implementation in how we use the vendor specific interface.

thoh-ot reviewed (approved):
LGTM
